apps/engine/lib/engine/transaction.ex

```diff
@@ -0,0 +1,91 @@
+defmodule Engine.Transaction do
+  @moduledoc """
+  Transaction model.
+  """
+
+  use Ecto.Schema
+  import Ecto.Changeset
+  import Ecto.Query, only: [from: 2]
+
+  @default_metadata <<0::160>>
+
+  @error_messages [
+    cannot_be_zero: "can't be zero",
+    exceeds_maximum: "can't exceed maximum value"
+  ]
+
+  schema "transactions" do
+    field(:tx_type, :integer, default: 1)
+    field(:tx_data, :integer, default: 0)
+    field(:metadata, :binary, default: @default_metadata)
+
+    belongs_to(:block, Engine.Block)
+    has_many(:inputs, Engine.Utxo, foreign_key: :spending_transaction_id)
+    has_many(:outputs, Engine.Utxo, foreign_key: :creating_transaction_id)
+
+    timestamps(type: :utc_datetime)
+  end
+
+  def insert(params) do
+    %__MODULE__{} |> changeset(params) |> Engine.Repo.insert()
+  end
+
+  defp changeset(struct, %{} = params) do
+    struct
+    |> Engine.Repo.preload(:inputs)
+    |> Engine.Repo.preload(:outputs)
+    |> cast(params, [:tx_type, :tx_data, :metadata])
+    |> validate_required([:tx_type, :tx_data, :metadata])
+    |> cast_assoc(:inputs)
+    |> cast_assoc(:outputs)
+    |> validate_usable_inputs()
+  end
+
+  defp changeset(struct, txbytes) when is_binary(txbytes) do
+    case ExPlasma.decode(txbytes) do
+      {:ok, transaction} ->
+        changeset(struct, params_from_ex_plasma(transaction))
+
+      {:error, {field, message}} ->
+        struct |> changeset(%{}) |> add_error(field, @error_messages[message])
+    end
+  end
+
+  @doc """
+  Validates that the given changesets inputs are correct. To create a transaction with inputs:
+
+    * The utxo position for the input must exist.
+    * The utxo position for the input must not have been spent.
+  """
+  defp validate_usable_inputs(changeset) do
+    positions =
+      changeset
+      |> get_field(:inputs)
+      |> Enum.map(&ExPlasma.Utxo.pos/1)
+
+    query =
+      from(u in Engine.Utxo,
+        where: u.pos in ^positions and is_nil(u.spending_transaction_id),
+        limit: 4
+      )
+
+    result = Engine.Repo.all(query)
+
+    if length(positions) != length(result) do
```

Comment:
pardon if the question is noob in terms of Ecto, but which part cares (or will care) about not allowing 2 transactions to spend the same UTXO at a given instant? Is this going to be taken care of on the `Engine.State` level?

Reply:
> *tl;dr* This is a first pass so it isn't included in this PR yet.

Great question. It's partially `Engine.Block` and a `Engine.BlockQueue`. This isn't specified in this PR yet but this will fall onto the `BlockQueue` responsibility and connects with how the `Utxo` exist in a two-phase state in the table:

* Initial Phase: utxo is created as an output, it contains `output_guard`, `amount`, `currency`
* Submitted Phase: When a block is formed and ready to submitted, we fill in that information with the `blknum`, `txindex`, and `oindex`(derived from ordering the transactions into said block)

The plan here is that a `BlockQueue` service will grab a `pending` block and submit them. if successful, it will update the associated `utxo` records with the second-half information (blknum, txindex, oindex) thereby forming verified, valid utxos. If the `BlockQueue` sees Utxos that are spent, it will mark the status as `spent`.

Successive transactions based on these utxos can use this method to check if the given utxo exists and are [unspent](https://github.com/omisego/elixir-omg/pull/1413/files/98e0f18e0294f7bc56ce023a60811524a76b8614#diff-c06fea0120d7565920d78d9fb89e977eR8).

Reply:
ah, ok, so you're essentially writing everything from scratch here :joy:

Then, the only thing that I can add here is that this seems to take away the possibility to "chain" txs in a single block, which comes at a risk and is discouraged (MoreVP specs) but sometimes leveraged (some dex designs, plasma dog) esp. for low-stake UTXOs. So you need to be aware of this, if you're taking this away.